README.md
# Metarouter Analytics

analytics-ios is an iOS client for Metarouter.

Special thanks to [Tony Xiao](https://github.com/tonyxiao), [Lee Hasiuk](https://github.com/lhasiuk) and [Cristian Bica](https://github.com/cristianbica) for their contributions to the library!

<div align="center">
  <img src="https://user-images.githubusercontent.com/1385202/73848360-a6a9d700-4830-11ea-929a-3394d4b0d2cf.png"/>
  <p><b><i>You can't fix what you can't measure</i></b></p>
</div>

Analytics helps you measure your users, product, and business. It unlocks insights into your app's funnel, core business metrics, and whether you have product-market fit.

## How to get started

[Metarouter](https://metarouter.io) collects analytics data and allows you to send it to more than 250 apps (such as Google Analytics, Mixpanel, Optimizely, Facebook Ads, Slack, Sentry) just by flipping a switch. You only need one Metarouter code snippet, and you can turn integrations on and off at will, with no additional code. [Sign up with Metarouter today](https://app.metarouter.io/signup).

### Why?
1. **Power all your analytics apps with the same data**. Instead of writing code to integrate all of your tools individually, send data to Metarouter, once.

2. **Install tracking for the last time**. We're the last integration you'll ever need to write. You only need to instrument Metarouter once. Reduce all of your tracking code and advertising tags into a single set of API calls.

3. **Send data from anywhere**. Send Metarouter data from any device, and we'll transform and send it on to any tool.

4. **Query your data in SQL**. Slice, dice, and analyze your data in detail with Metarouter SQL. We'll transform and load your customer behavioral data directly from your apps into Amazon Redshift, Google BigQuery, or Postgres. Save weeks of engineering time by not having to invent your own data warehouse and ETL pipeline.

    For example, you can capture data on any app:
    ```js
    analytics.track('Order Completed', { price: 99.84 })
    ```
    Then, query the resulting data in SQL:
    ```sql
    select * from app.order_completed
    order by price desc
    ```

## Installation

Analytics is available through [CocoaPods](http://cocoapods.org) and [Carthage](https://github.com/Carthage/Carthage).

### CocoaPods

```ruby
pod "MetarouterAnalytics", "~> 4.1.0"
```
Note: Metarouter _strongly_ recommends that you use a dynamic framework to manage your project dependencies. If you prefer static libraries, you can add `use_modular_headers!` or `use_frameworks! :linkage => :static` in your Podfile. However, you must then _manually update_ all of your dependencies on a regular schedule.

### Carthage

```
github "metarouter/analytics-ios"
```

### Swift Package Manager (SPM)

To add analytics-ios via Swift Package Mangaer, it is possible to add it one of two ways:

#### Xcode
![Xcode Add SPM Package](https://user-images.githubusercontent.com/1385202/124290377-073f5f80-db5c-11eb-89c1-04169a501097.jpg)

![ChoosePackageRepository](https://user-images.githubusercontent.com/1385202/124290416-13c3b800-db5c-11eb-94d2-0b017500a236.jpg)

![ChoosePackageOptions](https://user-images.githubusercontent.com/1385202/124290468-21793d80-db5c-11eb-907b-6395b5456cfc.jpg)

#### Package.swift
```
import PackageDescription

let package = Package(
    name: "MyApplication",
    dependencies: [
        // Add a package containing Analytics as the name along with the git url
        .package(
            name: "Metarouter",
            url: "git@github.com:metarouter/analytics-ios.git"
        )
    ],
    targets: [
        name: "MyApplication",
        dependencies: ["Metarouter"] // Add Analytics as a dependency of your application
    ]
)
```
Note: Metarouter recommends that you use Xcode to add your package.

## Quickstart

Refer to the Quickstart documentation at [https://docs.metarouter.io/docs/quick-start-guide](https://docs.metarouter.io/docs/quick-start-guide).

## Documentation

More detailed documentation is available at [https://docs.metarouter.io/docs/analyticsjs-for-ios](https://docs.metarouter.io/docs/analyticsjs-for-ios).

CHANGELOG.md
Change Log
==========
Version 4.2.0 *(22 December, 2021)*
-----------------------------
* [Fix] Removed unnecessary renames so syncing with upstream is simpler

Version 4.1.9 *(2 November, 2021)*
-----------------------------
* [New] Updated codebase with upstream (4.1.6 version)

Version 4.1.8 *(9 July, 2021)*
-----------------------------
* [New] Updated codebase with upstream

Version 4.1.5 *(9 July, 2021)*
-----------------------------
* [Fix](https://github.com/segmentio/analytics-ios/pull/999) Add missing nullability identifier 
* [Fix](https://github.com/segmentio/analytics-ios/pull/1002) Update Readme with SPM guide 

Version 4.1.4 *(3 May, 2021)*
-----------------------------
* [Fix](https://github.com/segmentio/analytics-ios/pull/994) Add missing import in Segment.h for Xcode 12.5
* [Fix](https://github.com/segmentio/analytics-ios/pull/995) Revert static lib podspec change.

Version 4.1.3 *(27 January, 2021)*
-----------------------------
* [Fix](https://github.com/segmentio/analytics-ios/pull/982) Pull IDFA if it has changed and rebuild context if necessary.
* [Fix](https://github.com/segmentio/analytics-ios/pull/973) Support Data Residency.
* [Fix](https://github.com/segmentio/analytics-ios/pull/968) Added static lib support for Cocoapods.

Version 4.1.2 *(12 November, 2020)*
-----------------------------
* [Fix](https://github.com/segmentio/analytics-ios/pull/966) Standarized all SDK versions between dependency managers. Migrated to importing modules vs headers. Removed all linked frameworks.
* [Fix](https://github.com/segmentio/analytics-ios/pull/963) Fixed missing userId in traits.
* [Fix](https://github.com/segmentio/analytics-ios/pull/962) Fixed an issue with NSCopying preceding SEGSerialization.
* [Fix](https://github.com/segmentio/analytics-ios/pull/957) Output log messages when events are trimmed from the queue.
* [Fix](https://github.com/segmentio/analytics-ios/pull/956) Added category option to screen calls.

Version 4.1.1 *(27 October, 2020)*
-----------------------------
* [Fix](https://github.com/segmentio/analytics-ios/pull/955) Fixed example projects using old branch for SPM.
* [Fix](https://github.com/segmentio/analytics-ios/pull/953) Repaired mac example project.
* [Fix] Fixed carthage support for main library and destinations.

Version 4.1.0 *(19 October, 2020)*
-----------------------------
* [New](https://github.com/segmentio/analytics-ios/pull/945) Added Webhook Integration.
* [New](https://github.com/segmentio/analytics-ios/pull/942) Renamed module from `Analytics` to `Segment`.
* [Fix](https://github.com/segmentio/analytics-ios/pull/941) Removed unused mobile attribution.
* [New](https://github.com/segmentio/analytics-ios/pull/940) Added ability for customer to make their types serializable.
* [New](https://github.com/segmentio/analytics-ios/pull/950) Added necessary changes for Edge Functions.

Version 4.0.5 (pre-release) *(28 August, 2020)*
-----------------------------
* [Fix](https://github.com/segmentio/analytics-ios/pull/937) Fixed issue where non-serializable types would get into payload.
* [Fix](https://github.com/segmentio/analytics-ios/pull/931) Fixed double device payload.
* [New](https://github.com/segmentio/analytics-ios/pull/930) Deprecated configuration access and updated associated tests.
* [Fix](https://github.com/segmentio/analytics-ios/pull/925) Disregard payloads that fail conversion from plist to json.
* [New](https://github.com/segmentio/analytics-ios/pull/922) Added Swift and Carthage badging.

Version 4.0.4 (pre-release) *(14 July, 2020)*
-----------------------------
* [New](https://github.com/segmentio/analytics-ios/pull/918) Added MacOS support.
* [Fix](https://github.com/segmentio/analytics-ios/pull/921) Fixed SEGLog permissions.
* [Fix](https://github.com/segmentio/analytics-ios/pull/920) Adopted Swift 5 syntax in tests. Removed SwiftTryCatch dependency for tests. Fixed deadlock from #916.

Version 4.0.3 (pre-release) *(6 July, 2020)*
-----------------------------
* [Fix](https://github.com/segmentio/analytics-ios/pull/917) Fixed an issue with old traits overwriting new traits.
* [New](https://github.com/segmentio/analytics-ios/pull/915) Added full example flow with Swift Package Manager compatibility.
* [Fix](https://github.com/segmentio/analytics-ios/pull/914) Fix SwiftUI previews in macOS Catalyst builds.

Version 4.0.2 (pre-release) *(25 June, 2020)*
-----------------------------
* [New](https://github.com/segmentio/analytics-ios/pull/913) Separate public utils from private utils appropriately.
* [New](https://github.com/segmentio/analytics-ios/pull/912) Payload Info & Traits Fixes.
* [New](https://github.com/segmentio/analytics-ios/pull/911) Updated SPM for code coverage compatibility.
* [New](https://github.com/segmentio/analytics-ios/pull/909) Removed Quick and Nimble, updated unit tests.

Version 4.0.1 (pre-release) *(18 June, 2020)*
-----------------------------
 * [New](https://github.com/segmentio/analytics-ios/pull/905) Added global state management; Middleware now receive all context data.
 * [New](https://github.com/segmentio/analytics-ios/pull/885) change logic for screen naming conventions.
 * [New](https://github.com/segmentio/analytics-ios/pull/899) Fixed crash on UISceneDelegate applications.
 * [New](https://github.com/segmentio/analytics-ios/pull/896) Renamed middleware methods to match a.js naming.
 * [New](https://github.com/segmentio/analytics-ios/pull/904) Added Swift Package Manager support.
 * [New](https://github.com/segmentio/analytics-ios/pull/902) Updated class names for Swift.

Version 4.0.0-beta.1 *(21st May, 2020)*
---------------------------------------
 * [New](https://github.com/segmentio/analytics-ios/pull/892) Makes IDFA support externally/customer driven.

Version 4.0.0-beta.0 *(22nd April, 2020)*
-----------------------------------------
 * [New](https://github.com/segmentio/analytics-ios/pull/879) Added integration middleware functionality.
 
Version 3.9.0 *(5th June, 2020)*

MetarouterTests/HTTPClientTest.swift
//
//  HTTPClientTest.swift
//  Analytics
//
//  Created by Tony Xiao on 9/16/16.
//  Copyright © 2016 Segment. All rights reserved.
//

// TODO: Uncomment these tests and get rid of Nocilla.

/*
import Nocilla
import Analytics
import XCTest

class HTTPClientTest: XCTestCase {
    
    var client: HTTPClient!
    let batch: [String: Any] = ["sentAt":"2016-07-19'T'19:25:06Z", "batch":[["type":"track", "event":"foo"]]]
    let context: [String: Any] = [
        "os": [
            "name": "iPhone OS",
            "version" : "8.1.3",
        ],
        "ip": "8.8.8.8",
    ]
    
    override func setUp() {
        super.setUp()
        LSNocilla.sharedInstance().start()
        client = HTTPClient(requestFactory: nil)
    }
    
    override func tearDown() {
        super.tearDown()
        LSNocilla.sharedInstance().clearStubs()
        LSNocilla.sharedInstance().stop()
    }
    
    func testDefaultRequestFactor() {
        let factory = HTTPClient.defaultRequestFactory()
        let url = URL(string: "https://e.metarouter.io/v1/batch")
        let request = factory(url!)
        XCTAssertEqual(request.url, url, "URLs should be the same")
    }
    
    func testSettingsForWriteKeySucceeds2xx() {
        _ = stubRequest("GET", "https://cdn-settings.segment.com/v1/projects/foo/settings" as NSString)
            .withHeader("User-Agent", "analytics-ios/" + Analytics.version())!
            .withHeaders(["Accept-Encoding" : "gzip" ])!
            .andReturn(200)!
            .withHeaders(["Content-Type" : "application/json"])!
            .withBody("{\"integrations\":{\"Segment.io\":{\"apiKey\":\"foo\"}},\"plan\":{\"track\":{}}}" as NSString)
        
        let doneExpectation = expectation(description: "Done with url session task")
        
        _ = client.settings(forWriteKey: "foo", completionHandler: { success, settings in
            
            XCTAssert(success, "Should be successful")
            XCTAssertEqual(settings as NSDictionary?, [
                "integrations": [
                    "Segment.io": [
                        "apiKey":"foo"
                    ]
                ],
                "plan":[
                    "track": [:]
                ]
                ] as NSDictionary)
            doneExpectation.fulfill()
        })
        
        wait(for: [doneExpectation], timeout: 1.0)
    }
    
    func testSettingsWriteKey2xxResponse() {
        _ = stubRequest("GET", "https://cdn-settings.segment.com/v1/projects/foo/settings" as NSString)
            .withHeader("User-Agent", "analytics-ios/" + Analytics.version())!
            .withHeaders(["Accept-Encoding" : "gzip" ])!
            .andReturn(400)!
            .withHeaders(["Content-Type" : "application/json" ])!
            .withBody("{\"integrations\":{\"Segment.io\":{\"apiKey\":\"foo\"}},\"plan\":{\"track\":{}}}" as NSString)
        
        let doneExpectation = expectation(description: "Done with url session task")
        
        client.settings(forWriteKey: "foo", completionHandler: { success, settings in
            XCTAssertFalse(success, "Success should be false")
            XCTAssertNil(settings, "Failure should have nil settings")
            
            doneExpectation.fulfill()
        })
        
        wait(for: [doneExpectation], timeout: 1.0)
        
    }
    
    func testSettingsWriteKey2xxJSONErrorResponse() {
        _ = stubRequest("GET", "https://cdn-settings.segment.com/v1/projects/foo/settings" as NSString)
            .withHeader("User-Agent", "analytics-ios/" + Analytics.version())!
            .withHeaders(["Accept-Encoding":"gzip"])!
            .andReturn(200)!
            .withHeaders(["Content-Type":"application/json"])!
            .withBody("{\"integrations" as NSString)
        
        let doneExpectation = expectation(description: "Done with url session task")
        
        client.settings(forWriteKey: "foo", completionHandler: { success, settings in
            XCTAssertFalse(success, "Success should be false")
            XCTAssertNil(settings, "Failure should have nil settings")
            doneExpectation.fulfill()
        })
        wait(for: [doneExpectation], timeout: 1.0)
    }
    
    func testUploadNoRetry() {
        let batch: [String: Any] = [
            // Dates cannot be serialized as is so the json serialzation will fail.
            "sentAt": NSDate(),
            "batch": [["type": "track", "event": "foo"]],
        ]
        let doneExpectation = expectation(description: "Done with url session task")
        let task = client.upload(batch, forWriteKey: "bar") { retry in
            XCTAssertFalse(retry, "Retry should be false")
            doneExpectation.fulfill()
        }
        XCTAssertNil(task, "Task should be nil")
        wait(for: [doneExpectation], timeout: 1.0)
    }
    
    func testUploadNoRetry2xx() {
        _ = stubRequest("POST", "https://e.metarouter.io/v1/batch" as NSString)
            .withHeader("User-Agent", "analytics-ios/" + Analytics.version())!
            .withJsonGzippedBody(batch as AnyObject)
            .withWriteKey("bar")
            .andReturn(200)
        let doneExpectation = expectation(description: "Done with url session task")
        _ = client.upload(batch, forWriteKey: "bar") { retry in
            XCTAssertFalse(retry, "Retry should be false")
            doneExpectation.fulfill()
        }
        wait(for: [doneExpectation], timeout: 1.0)
    }
    
    func testUploadRetry3xx() {
        _ = stubRequest("POST", "https://e.metarouter.io/v1/batch" as NSString)
            .withHeader("User-Agent", "analytics-ios/" + Analytics.version())!
            .withJsonGzippedBody(batch as AnyObject)
            .withWriteKey("bar")
            .andReturn(304)
        let doneExpectation = expectation(description: "Done with url session task")
        _ = client.upload(batch, forWriteKey: "bar") { retry in
            XCTAssert(retry, "Retry should be true")
            doneExpectation.fulfill()
        }
        wait(for: [doneExpectation], timeout: 1.0)
    }
    
    func testUploadNoRetry4xx() {
        _ = stubRequest("POST", "https://e.metarouter.io/v1/batch" as NSString)
            .withHeader("User-Agent", "analytics-ios/" + Analytics.version())!
            .withJsonGzippedBody(batch as AnyObject)
            .withWriteKey("bar")
            .andReturn(401)
        let doneExpectation = expectation(description: "Done with url session task")
        _ = client.upload(batch, forWriteKey: "bar") { retry in
            XCTAssertFalse(retry, "Retry should be false")
            doneExpectation.fulfill()
        }
        wait(for: [doneExpectation], timeout: 1.0)
    }
    
    func testRetryFor429() {
        _ = stubRequest("POST", "https://e.metarouter.io/v1/batch" as NSString)
            .withHeader("User-Agent", "analytics-ios/" + Analytics.version())!
            .withJsonGzippedBody(batch as AnyObject)
            .withWriteKey("bar")
            .andReturn(429)
        let doneExpectation = expectation(description: "Done with url session task")
        _ = client.upload(batch, forWriteKey: "bar") { retry in
            XCTAssert(retry, "Retry should be true")
            doneExpectation.fulfill()
        }
        wait(for: [doneExpectation], timeout: 1.0)
    }
    
    func testRetryFor5xx() {
        _ = stubRequest("POST", "https://e.metarouter.io/v1/batch" as NSString)
            .withHeader("User-Agent", "analytics-ios/" + Analytics.version())!
            .withJsonGzippedBody(batch as AnyObject)
            .withWriteKey("bar")
            .andReturn(504)
        let doneExpectation = expectation(description: "Done with url session task")
        _ = client.upload(batch, forWriteKey: "bar") { retry in
            XCTAssert(retry, "Retry should be true")
            doneExpectation.fulfill()
        }
        wait(for: [doneExpectation], timeout: 1.0)
    }
    
    func testBatchSizeFailure() {
        let oversizedBatch: [String: Any] = ["sentAt":"2016-07-19'T'19:25:06Z",
                                             "batch": Array(repeating: ["type":"track", "event":"foo"], count: 16000)]
        let doneExpectation = expectation(description: "Done with url session task")
        _ = client.upload(oversizedBatch, forWriteKey: "bar") { retry in
            XCTAssertFalse(retry, "Retry should be false")
            doneExpectation.fulfill()
        }
        wait(for: [doneExpectation], timeout: 1.0)
    }
}
*/

